alf/algorithms/muzero_representation_learner.py

```diff
@@ -930,3 +931,160 @@ def __call__(self, age, max_td_steps, current_max_age):
+        updated.mini_batch_size = training_options.mini_batch_size
+        updated.num_updates_per_train_iter = training_options.num_updates_per_train_iter
+        updated.replay_buffer_length = training_options.replay_buffer_length
+        updated.initial_collect_steps = training_options.initial_collect_steps
+        updated.priority_replay = training_options.priority_replay
+        updated.priority_replay_alpha = as_scheduler(
+            training_options.priority_replay_alpha)
+        updated.priority_replay_beta = as_scheduler(
+            training_options.priority_replay_beta)
+
+        self._impl = impl_cls(
+            observation_spec=observation_spec,
+            action_spec=action_spec,
+            reward_spec=reward_spec,
+            config=updated,
+            debug_summaries=debug_summaries)
+        self._impl.force_params_visible_to_parent = True
+        assert self._impl._reanalyze_ratio in [
+            0.0, 1.0
+        ], ('Currently MuzeroRepresentationLearner only support reanalyze ratio 0.0 or 1.0'
+            )
+
+        if self._impl._reanalyze_ratio > 0:
+            assert config.use_rollout_state, (
+                'use_rollout_state needs to be True when reanalyze is used.')
+
+    @property
+    def output_spec(self):
+        """Access the spec of the produced representation.
+
+        This will be used as the obervation spec for the subsequent RL
+        algorithm.
+
+        """
+        return self._impl._model.repr_spec
+
+    def rollout_step(self, time_step: TimeStep, state):
+        repr_step = self._impl.rollout_step(time_step, state)
+        # Save in the representation learner's own replay buffer. Note that
+        # ``observe_for_replay`` when called for the first time will have the
+        # side effect of creating the replay buffer.
+        if self._impl._replay_buffer is None:
+            self._impl.set_replay_buffer(
+                time_step.env_id.shape[0],
+                self._training_options.replay_buffer_length,
+                self._training_options.priority_replay)
+        exp = make_experience(time_step.untransformed, repr_step, state)
+        self._impl.observe_for_replay(exp)
+        return repr_step
+
+    def train_step(self, exp: TimeStep, state, rollout_info):
+        return self._impl.predict_step(exp, state)
+
+    def preprocess_experience(self, root_inputs: TimeStep, rollout_info,
+                              batch_info):
+        return root_inputs, ()
+
+    def calc_loss(self, info):
+        # The calc_loss() here does nothing so that ``Agent`` will only handle
+        # the loss from other sub algorithm such as RL algorithm.
+        #
+        # The actual loss for training the representation itself is within
+        # ``self._impl``.
+        return LossInfo(loss=(), extra={})
+
+    def after_update(self, root_inputs, info):
+        pass
+
+    def after_train_iter(self, experience, info):
+        if self._training_options is None:
```

Comment:
It seems that the current implementation does not really support `training_options == None` ?
But it might be useful to support this to use it for algorithms such as SAC.

Reply:
Right. Updated the logic so that `training_options == None` means not training representation.